runners/spark/src/main/java/org/apache/beam/runners/spark/io/MicrobatchSource.java

```diff
@@ -321,6 +318,21 @@ public Instant getWatermark() {
     }
   }
 
+  /**
+   * Listener to be called when a reader is removed from {@link MicrobatchSource#readerCache}.
+   */
+  private static class ReaderCacheRemovalListener
+      implements RemovalListener<MicrobatchSource<?, ?>, BoundedReader<?>> {
+    @Override public void onRemoval(
+        RemovalNotification<MicrobatchSource<?, ?>, BoundedReader<?>> notification) {
+      try {
+        notification.getValue().close();
```

Comment:
Is there a scenario where the cache gets evicted and immediately populated again with the same key ? if so, is the cache thread-safe ? will the `notification` hold reference to the removed value, or the newly populated value ?